feat: add beagle flutter documentation (#717)
ZupIT/beagle-docs: content/en/resources/customization/beagle-for-flutter/customization/actions.md

---
title: Custom Actions
weight: 3
description: >-

>   You will find here information of how to create and use your own actions in Beagle Flutter.

Reviewer: ```suggestion
In this section, you will find information on how to create and use your own actions in Beagle Flutter.
```

---

---

## How to create custom actions?

> Just like custom componentes, it's pretty simple to create a custom actions in Beagle Flutter, you need to declare a map that follows the strcut `Map<String, ActionHandler>` and pass it to the initialization method `BeagleSdk.init`. Basically this struct tells Beagle which action trigger when a given identifier comes into JSON.

Reviewer: ```suggestion
It's simple to create custom actions in Beagle Flutter, you need to: 
- Declare a map that follows the struct `Map<String, ActionHandler>`; 
- Pass it to the initialization method `BeagleSdk.init`. 

This struct tells Beagle which action trigger when a given identifier comes into JSON.
```

> Just like custom componentes, it's pretty simple to create a custom actions in Beagle Flutter, you need to declare a map that follows the strcut `Map<String, ActionHandler>` and pass it to the initialization method `BeagleSdk.init`. Basically this struct tells Beagle which action trigger when a given identifier comes into JSON.

Reviewer 2: ```suggestion
Just like custom componentes, it's pretty simple to create a custom action in Beagle Flutter. You need to declare a map that follows the struct `Map<String, ActionHandler>` and pass it to the initialization method `BeagleSdk.init`. Basically, this struct tells Beagle which action to trigger when a given identifier comes from a 
 JSON.
```

### String key definition
The String key of the map is the `_beagleAction_` property in JSON that identifies the action type. You can name it whatever you like, as long as the prefix is ​​`custom:`. 

### ActionHandler value definition

> The ActionHandler value of the map is a function that receives a `BeagleAction`, a `BeagleView`, a `BeagleUIElement` and a context as parameters, following the strcut `void Function({BeagleAction action, BeagleView view, BeagleUIElement element, BuildContext context})`. This parameters are available because they provide all the information needed to trigger an event, such as context and the action coming from JSON.

Reviewer: ```suggestion
The ActionHandler value of the map is a function that receives a:
1. `BeagleAction`; 
2. `BeagleView`;
3. `BeagleUIElement`;
4.  A context as parameters following the struct `void Function({BeagleAction action, BeagleView view, BeagleUIElement element, BuildContext context})`. 

These parameters are available because they provide all the information needed to trigger an event, such as context and the action coming from JSON.
```

## Example

> Follow this steps to learn how to create and use your own custom actions:

Reviewer: ```suggestion
Follow these steps to learn how to create and use your own custom actions:
```

### Step 1: Create the action identifier

> The first step to create a custom action with Beagle Flutter, is to define the identifier. Here, let's name it `custom:log`.

Reviewer: ```suggestion
You have to define the identifier. Here, let's name it `custom:log`.
```

> The first step to create a custom action with Beagle Flutter, is to define the identifier. Here, let's name it `custom:log`.

Reviewer 2: ```suggestion
The first step to create a custom action with Beagle Flutter is to define the identifier. Here, let's name it `custom:log`.
```

### Step 2: Create the action

> Now create your action regardless of Beagle. For this example, we'll use the simple function below:

Reviewer: ```suggestion
Create your action regardless Beagle. For this example we used a simple function, see below:
```

> Now create your action regardless of Beagle. For this example, we'll use the simple function below:

Reviewer 2: regardless of Beagle **what**? action, implementation, interface, lib... ?

```dart
void debug(String message) {
  debugPrint(message);
}
```

### Step 3: Create the custom actions map

> With the function and action identifier done, it's time to create the struct that we'll pass to Beagle:

Reviewer: ```suggestion
Create the struct that it will pass to Beagle:
```

```dart
Map<String, ActionHandler> myCustomActions = {
  'custom:log': ({action, view, element, context}) {
    debug(action.getAttributeValue('message'));
  }
};
```

> ### Step 4: Pass them to Beagle

Reviewer 2: ```suggestion
### Step 4: Pass the custom map to Beagle
```

> To Beagle be able to trigger your action, pass the map you just created through the Beagle init method:

Reviewer: ```suggestion
Beagle needs to be able to trigger your action, you have to pass the map you created through the Beagle init method:
```

> To Beagle be able to trigger your action, pass the map you just created through the Beagle init method:

Reviewer 2: ```suggestion
For Beagle to be able to trigger your action, pass the map you just created through the Beagle init method:
```
